zcash_client_backend/src/data_api/testing/pool.rs

```diff
@@ -450,6 +450,1704 @@ pub fn zip_315_confirmations_test_steps<T: ShieldedPoolTester>(
+
+    // Now, store the transaction, pretending it has been mined (we will actually mine the block
+    // next). This will cause the the gap start to move & a new `gap_limits.ephemeral()` of
+    // addresses to be created.
+    let target_height = st.latest_cached_block().unwrap().height() + 1;
+    st.wallet_mut()
+        .store_decrypted_tx(DecryptedTransaction::new(
+            Some(target_height),
+            build_result.transaction(),
+            vec![],
+            #[cfg(feature = "orchard")]
+            vec![],
+        ))
+        .unwrap();
+
+    // Mine the transaction & scan it so that it is will be detected as mined. Note that
+    // `generate_next_block_including` does not actually do anything with fully-transparent
+    // transactions; we're doing this just to get the mined block that we added via
+    // `store_decrypted_tx` into the database.
+    let (h, _) = st.generate_next_block_including(txid);
+    st.scan_cached_blocks(h, 1);
+    assert_eq!(h, target_height);
+
+    // At this point the start of the gap should be at index `gap_limits.ephemeral()` and the new
+    // size of the known address set should be `gap_limits.ephemeral() * 2`.
+    let new_known_addrs = st
+        .wallet()
+        .get_known_ephemeral_addresses(account_id, None)
+        .unwrap();
+    assert_eq!(
+        new_known_addrs.len(),
+        usize::try_from(gap_limits.ephemeral() * 2).unwrap()
+    );
+    assert!(new_known_addrs.starts_with(&known_addrs));
+
+    let reservation_should_succeed = |st: &mut TestState<_, DSF::DataStore, _>, n: u32| {
+        let reserved = st
+            .wallet_mut()
+            .reserve_next_n_ephemeral_addresses(account_id, n.try_into().unwrap())
+            .unwrap();
+        assert_eq!(reserved.len(), usize::try_from(n).unwrap());
+        reserved
+    };
+    let reservation_should_fail =
+        |st: &mut TestState<_, DSF::DataStore, _>, n: u32, expected_bad_index| {
+            assert_matches!(st
+            .wallet_mut()
+            .reserve_next_n_ephemeral_addresses(account_id, n.try_into().unwrap()),
+            Err(e) if is_reached_gap_limit(&e, account_id, expected_bad_index));
+        };
+
+    let next_reserved = reservation_should_succeed(&mut st, 1);
+    assert_eq!(
+        next_reserved[0],
+        known_addrs[usize::try_from(gap_limits.ephemeral()).unwrap()]
+    );
+
+    // The range of address indices that are safe to reserve now is
+    // 0..(gap_limits.ephemeral() * 2 - 1)`, and we have already reserved or used
+    // `gap_limits.ephemeral() + 1`, addresses, so trying to reserve another
+    // `gap_limits.ephemeral()` should fail.
+    reservation_should_fail(&mut st, gap_limits.ephemeral(), gap_limits.ephemeral() * 2);
+    reservation_should_succeed(&mut st, gap_limits.ephemeral() - 1);
+    // Now we've reserved everything we can, we can't reserve one more
+    reservation_should_fail(&mut st, 1, gap_limits.ephemeral() * 2);
```

Comment:
@pacu why is this test checking all this ephemeral address gap limit stuff? It shouldn't be needed for what we want to test here.

Reply:
oh I just replicated the send to text but with max

Reply:
Removed in https://github.com/zcash/librustzcash/pull/1733/commits/0ad99b2815d605dc0537af102214f5600b7caacf